zapisy/apps/schedule/models/term.py

```diff
@@ -210,59 +211,118 @@ def __str__(self):
-        return
-    if not instance.pk:
-        return
-    instance = CourseTerm.objects.get(pk=instance.pk)
-    dates = instance.group.course.semester.get_all_days_of_week(instance.dayOfWeek)
-    matching_terms = Term.objects.filter(event__group=instance.group,
+def get_terms(course_term: CourseTerm):
+    dates = course_term.group.course.semester.get_all_days_of_week(course_term.dayOfWeek)
+    matching_terms = Term.objects.filter(event__group=course_term.group,
                                          day__in=dates,
-                                         start=instance.start_time,
-                                         end=instance.end_time)
+                                         start=course_term.start_time,
+                                         end=course_term.end_time)
+    return matching_terms
+
+
+def delete_terms(course_term: CourseTerm, room_pk_set: Optional[Set[int]] = None):
+    """Deletes all Terms corresponding to a CourseTerm."""
+    matching_terms = get_terms(course_term)
+    if room_pk_set:
+        matching_terms = matching_terms.filter(room__pk__in=room_pk_set)
     matching_terms.delete()
 
 
-@receiver(models.signals.post_save, sender=CourseTerm)
-@receiver(models.signals.m2m_changed, sender=CourseTerm.classrooms.through)
-def create_course_terms(**kwargs):
-    """Creates matching Terms when a new CourseTerm is created."""
-    if not kwargs.get('action', 'post_save').startswith('post_'):
-        # We are in a post_save action of m2m_changed signal receiver.
-        # The function above is handling that.
-        return
-    instance: CourseTerm = kwargs['instance']
-    dates = instance.group.course.semester.get_all_days_of_week(instance.dayOfWeek)
-    event, _ = Event.objects.get_or_create(group=instance.group,
-                                           course=instance.group.course,
-                                           title=instance.group.course.get_short_name(),
+def create_terms(course_term: CourseTerm, room_pk_set: Optional[Set[int]] = None):
+    """Creates Terms corresponding to a CourseTerm."""
+    dates = course_term.group.course.semester.get_all_days_of_week(course_term.dayOfWeek)
+    event, _ = Event.objects.get_or_create(group=course_term.group,
+                                           course=course_term.group.course,
+                                           title=course_term.group.course.get_short_name(),
                                            type=Event.TYPE_CLASS,
                                            visible=True,
                                            status=Event.STATUS_ACCEPTED,
-                                           author=instance.group.teacher.user)
+                                           author=course_term.group.teacher.user)
+    if course_term.pk:
+        if room_pk_set:
+            rooms = Classroom.objects.filter(pk__in=room_pk_set)
+        else:
+            rooms = course_term.classrooms.all()
+    else:
+        rooms = None
```

Comment:
Zdaje się że jeśli zrobimy tak
```suggestion
        rooms = [None]
```
to w następnej pętli będzie można zdeduplikować jedną (ale długą!) instrukcję.